File: backends/ts-hono/src/routes/adminInvitations.ts
/**
 * 管理者招待ルーターモジュール。
 *
 * @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
 */
import { Hono } from "hono";
import { db, asTx } from "../db/client.js";
import { DrizzleInvitationRepository } from "../infrastructure/persistence/drizzleInvitationRepository.js";
import { RedisInvitationAuthRepository } from "../infrastructure/cache/redisInvitationAuthRepository.js";
import { InvitationInteractor } from "../usecase/invitation/interactor.js";
import { badRequest, unauthorized } from "../lib/errors.js";
import { getStaffIdFromCookie } from "../lib/cookie.js";

const invitationAuthRepo = new RedisInvitationAuthRepository();

const app = new Hono();

function parseRole(raw: string | undefined): number {
  const role = raw !== undefined ? parseInt(raw, 10) : 2;
  if (role !== 1 && role !== 2) throw badRequest("invalid_role");
  return role;
}

app.get("/invitation", async (c) => {
  const role = parseRole(c.req.query("role"));
  const uc = new InvitationInteractor(new DrizzleInvitationRepository(db), invitationAuthRepo);
  const result = await uc.current(role);
  return c.json({ found: true, url: result.url, display_url: result.displayUrl, token: result.token });
});

app.get("/invitation/issue", async (c) => {
  const staffId = getStaffIdFromCookie(c);
  if (!staffId) throw unauthorized("unauthenticated");
  const role = parseRole(c.req.query("role"));
  let result!: Awaited<ReturnType<InvitationInteractor["issue"]>>;
  await db.transaction(async (tx) => {
    const uc = new InvitationInteractor(new DrizzleInvitationRepository(asTx(tx)), invitationAuthRepo);
    result = await uc.issue(role);
  });
  return c.json({ found: true, url: result.url, display_url: result.displayUrl, token: result.token });
});

export default app;

File: backends/python-fastapi/app/routers/admin_invitations.py
"""
管理者招待ルーターモジュール。

Author: Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
"""
from fastapi import APIRouter, Depends, Query
from app.exceptions import bad_request, unauthorized
from app.routers.deps import get_invitation_interactor, get_staff_id_from_cookie
from app.usecase.invitation.interactor import InvitationInteractor

router = APIRouter()


@router.get("/admin/invitation")
def index(
    role: int = Query(2),
    interactor: InvitationInteractor = Depends(get_invitation_interactor),
):
    if role not in (1, 2):
        raise bad_request("invalid_role")
    result = interactor.current(role)
    return {"found": True, "url": result.url, "display_url": result.display_url, "token": result.token}


@router.get("/admin/invitation/issue")
def issue(
    role: int = Query(2),
    staff_id: int = Depends(get_staff_id_from_cookie),
    interactor: InvitationInteractor = Depends(get_invitation_interactor),
):
    if staff_id == 0:
        raise unauthorized("unauthenticated")
    if role not in (1, 2):
        raise bad_request("invalid_role")
    result = interactor.issue(role)
    return {"found": True, "url": result.url, "display_url": result.display_url, "token": result.token}
